### PR TITLE
FIM: Remove audit rules when monitoring symbolic links

### LINKED ISSUE #5761: Audit rules are not being removed after symbolic link changes its target
|Wazuh version|Branch|Component|Install type|Install method|Platform|
|---|---|---|---|---|---|
| 4.0.0 | develop | ossec-syscheck | Manager/Agent | Packages/Sources | Linux |

## Description

Using symbolic links, when changing its target with _whodata_ enabled, _audit_ rules are not being removed from the previously monitored file and FIM tries to scan not monitored files.

## Example
When changing the target from `/file_3` to `/test/file_1`, the link is updated correctly in FIM:
```
2020/08/18 12:14:11 ossec-syscheckd[102138] run_check.c:557 at symlink_checker_thread(): INFO: (6034): Updating the symbolic link from '/file_3': '/test/linko' to '/test/file_1'.
```

However, after waiting for _audit_ rules to reload, this message appears in the log:
```
2020/08/18 12:16:41 ossec-syscheckd[102138] syscheck_audit.c:1120 at audit_reload_rules(): DEBUG: (6275): Reloading Audit rules.
2020/08/18 12:16:41 ossec-syscheckd[102138] audit_op.c:74 at audit_print_reply(): DEBUG: Audit rule loaded: -w /file_3 -p wa -k wazuh_fim
2020/08/18 12:16:41 ossec-syscheckd[102138] audit_op.c:74 at audit_print_reply(): DEBUG: Audit rule loaded: -w /test/file_1 -p wa -k wazuh_fim
2020/08/18 12:16:41 ossec-syscheckd[102138] syscheck_audit.c:1122 at audit_reload_rules(): DEBUG: (6276): Audit rules reloaded. Rules loaded: 1
```

According to this message, both `/file_3` and `/test/file_1` have `audit` rules, but only `/test/file_1` should have rules as it is the file pointed to by the symbolic link.

This messages are written in the log every time the _audit_ rules are reloaded. Because the rule is not removed for `/file_3`, this kind of message will appear in the log:
```
2020/08/18 12:18:22 ossec-syscheckd[102138] create_db.c:668 at fim_configuration_directory(): DEBUG: (6319): No configuration found for (file):'/file_3'
```

The `No configuration found...` message is appearing also when monitoring a link using _realtime_, so the same issue affecting the _audit_ rules may be affecting the _inotify_ watches.

### PR DESCRIPTION
**Pull request by antoniomanuelfr**
|Related issue|
|---|
|#5761|

## Description
This PR aims to fix the problem described in #5761. When a symlink is updated or removed, the audit rule is removed.
Closes #5761 
## Configuration options
```xml
    <directories whodata="yes" follow_symbolic_link="yes">/test_link,/test_link2,/test_link3</directories>
```
## Logs/Alerts example

```
2020/12/23 15:55:09 wazuh-syscheckd[6764] run_check.c:579 at symlink_checker_thread(): DEBUG: (6296): Links check finalized.
2020/12/23 15:55:11 wazuh-syscheckd[6764] run_check.c:536 at symlink_checker_thread(): DEBUG: (6033): Starting symbolic link updater. Interval '2'.
2020/12/23 15:55:11 wazuh-syscheckd[6764] run_check.c:550 at symlink_checker_thread(): INFO: (6034): Updating symbolic link '/symlink': from '/testdir1' to '/testdir_not_target'.
2020/12/23 15:55:11 wazuh-syscheckd[6764] fim_db.c:380 at fim_db_check_transaction(): DEBUG: Database transaction completed.
2020/12/23 15:55:25 wazuh-syscheckd[6764] syscheck_audit.c:1161 at audit_reload_rules(): DEBUG: (6275): Reloading Audit rules.
2020/12/23 15:55:25 wazuh-syscheckd[6764] syscheck_audit.c:236 at add_audit_rules_syscheck(): DEBUG: (6270): Added audit rule for monitoring directory: '/testdir_not_target'
2020/12/23 15:55:25 wazuh-syscheckd[6764] syscheck_audit.c:1163 at audit_reload_rules(): DEBUG: (6276): Audit rules reloaded. Rules loaded: 1
2020/12/23 15:55:25 wazuh-syscheckd[6764] run_check.c:536 at symlink_checker_thread(): DEBUG: (6033): Starting symbolic link updater. Interval '2'.
```
The output of `auditctl -l` after updating de links: 
```
[root@centosmanager vagrant]# auditctl -l
-w /testdir_not_target -p wa -k wazuh_fim
[root@centosmanager vagrant]# 

```
## Tests

- Compilation without warnings in every supported platform
  - [X] Linux
  - [x] Windows
  - [ ] MAC OS X
- [X] Source installation
- [X] Source upgrade

- Memory tests for Linux
  - [x] Scan-build report
  - [x] Coverity
  - [x] Valgrind (memcheck and descriptor leaks check)

### CODE REVIEW

**Comment by antoniomanuelfr** [comment]
scan-build did not report any error: 
```
Done building server

scan-build: Analysis run complete.
scan-build: Removing directory '/tmp/scan-build-2020-12-23-164707-103237-1' because it contains no reports.
scan-build: No bugs found.
```

File: src/unit_tests/syscheckd/test_syscheck_audit.c [hunk 1]
```diff
@@ -2654,8 +2622,8 @@ int main(void) {
         cmocka_unit_test(test_audit_get_id_begin_error),
         cmocka_unit_test(test_audit_get_id_end_error),
         cmocka_unit_test(test_init_regex),
-        cmocka_unit_test_setup_teardown(test_add_audit_rules_syscheck_added, setup_add_audit_rules, teardown_add_audit_rules),
-        cmocka_unit_test_setup_teardown(test_add_audit_rules_syscheck_not_added, setup_add_audit_rules, teardown_add_audit_rules),
+        cmocka_unit_test_setup_teardown(test_add_audit_rules_syscheck_duplicate_entry, setup_add_audit_rules, teardown_add_audit_rules),
+        //cmocka_unit_test_setup_teardown(test_add_audit_rules_syscheck_not_added, setup_add_audit_rules, teardown_add_audit_rules),
```
Comment (jotacarma90) [inline src/unit_tests/syscheckd/test_syscheck_audit.c:2626]: This test is commented

**Review by jotacarma90 — CHANGES_REQUESTED** [review]

**Review by jotacarma90 — APPROVED** [review]

**Review by jotacarma90 — DISMISSED** [review]
LGTM

File: src/unit_tests/syscheckd/test_syscheck_audit.c [hunk 2]
```diff
@@ -2171,7 +2142,7 @@ void test_audit_health_check_fail_to_create_hc_file(void **state) {
 
     hc_thread_active = 0;
 
-    will_return(__wrap_audit_add_rule, -17);
+    will_return(__wrap_audit_add_rule, EEXIST);
```
Comment (Molter73) [inline src/unit_tests/syscheckd/test_syscheck_audit.c:2145]: Should this be `-EEXIST`?

File: src/unit_tests/syscheckd/test_syscheck_audit.c [hunk 3]
```diff
@@ -2208,7 +2179,7 @@ void test_audit_health_check_no_creation_event_detected(void **state) {
...
```
Comment (Molter73) [inline src/unit_tests/syscheckd/test_syscheck_audit.c:2182]: Should this be `-EEXIST`?

**Review by Molter73 — CHANGES_REQUESTED** [review]

**Comment by antoniomanuelfr** [comment]
This is the result of the execution of the integration test locally (tier 0 and tier 1):
[report_audit.zip](https://github.com/wazuh/wazuh/files/5841403/report_audit.zip)

**Comment by antoniomanuelfr** [comment]
These are the result of the execution of the integration tests in Jenkins:

[agent_centos_html_report_test_integration_B1025_20210120092604.zip](https://github.com/wazuh/wazuh/files/5841926/agent_centos_html_report_test_integration_B1025_20210120092604.zip)
[agent_ubuntu_html_report_test_integration_B1025_20210120092604.zip](https://github.com/wazuh/wazuh/files/5841927/agent_ubuntu_html_report_test_integration_B1025_20210120092604.zip)
[agent_windows_html_report_test_integration_B1025_20210120092604.zip](https://github.com/wazuh/wazuh/files/5841928/agent_windows_html_report_test_integration_B1025_20210120092604.zip)

**Review by Molter73 — CHANGES_REQUESTED** [review]

**Review by Molter73 — APPROVED** [review]
LGTM!

**Comment by antoniomanuelfr** [comment]
These are the result of the integration test after all the modifications: 

[agent_macos_html_report_test_integration_B1039_20210122132156.zip](https://github.com/wazuh/wazuh/files/5856510/agent_macos_html_report_test_integration_B1039_20210122132156.zip)
[agent_solaris_html_report_test_integration_B1039_20210122132156.zip](https://github.com/wazuh/wazuh/files/5856511/agent_solaris_html_report_test_integration_B1039_20210122132156.zip)

**Comment by antoniomanuelfr** [comment]
After the latest changes in this PR (rebase to master to fix conflicts with https://github.com/wazuh/wazuh/pull/6789) these are the results of the integration tests executed in Jenkins:
[agent_centos_html_report_test_integration_B1046_20210126162246.zip](https://github.com/wazuh/wazuh/files/5878310/agent_centos_html_report_test_integration_B1046_20210126162246.zip)
[agent_macos_html_report_test_integration_B1047_20210126162317.zip](https://github.com/wazuh/wazuh/files/5878311/agent_macos_html_report_test_integration_B1047_20210126162317.zip)
[agent_solaris_html_report_test_integration_B1047_20210126162317.zip](https://github.com/wazuh/wazuh/files/5878312/agent_solaris_html_report_test_integration_B1047_20210126162317.zip)
[agent_ubuntu_html_report_test_integration_B1046_20210126162246.zip](https://github.com/wazuh/wazuh/files/5878313/agent_ubuntu_html_report_test_integration_B1046_20210126162246.zip)
[agent_windows_html_report_test_integration_B1046_20210126162246.zip](https://github.com/wazuh/wazuh/files/5878314/agent_windows_html_report_test_integration_B1046_20210126162246.zip)
Kind regards

**Review by Molter73 — CHANGES_REQUESTED** [review]
I just tried to replace a monitored symbolic link with a directory and the rule for the target remains configured.

Monitoring:
```xml
<directories whodata="yes" follow_symbolic_link="yes">/home/vagrant/link</directories>
```

With `/home/vagrant/link -> /home/vagrant/monitored-wd`, I issued the following command:
```
vagrant@server:~$ rm link && mkdir link && touch link/test
```

After the links and audit rules are reloaded I get:
```
root@server:/home/vagrant# auditctl -l
-w /etc/group -p wa
-w /home/vagrant/monitored-wd -p wa -k wazuh_fim
-w /home/vagrant/link -p wa -k wazuh_fim
```

And if I stop `wazuh-syscheckd` only the rule pointing to `/home/vagrant/link` is removed.

I also found some memory leaks by running the daemon compiled with AddressSanitizer:
```
=================================================================
==5498==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 82 byte(s) in 12 object(s) allocated from:
    #0 0x7fc319f74538 in strdup (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x77538)
    #1 0x5612d9bb9ff6 in fim_get_real_path syscheckd/create_db.c:1239
    #2 0x5612d9bc666b in fim_link_update syscheckd/run_check.c:600
    #3 0x5612d9bc62f5 in symlink_checker_thread syscheckd/run_check.c:553
    #4 0x7fc3193706da in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76da)

Direct leak of 27 byte(s) in 1 object(s) allocated from:
    #0 0x7fc319f74538 in strdup (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x77538)
    #1 0x5612d9bc8353 in fim_link_reload_broken_link syscheckd/run_check.c:765
    #2 0x5612d9bc651f in symlink_checker_thread syscheckd/run_check.c:573
    #3 0x7fc3193706da in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76da)

Direct leak of 27 byte(s) in 1 object(s) allocated from:
    #0 0x7fc319f74538 in strdup (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x77538)
    #1 0x5612d9c2347c in dump_syscheck_file config/syscheck-config.c:213
    #2 0x5612d9c28aae in read_attr config/syscheck-config.c:1282
    #3 0x5612d9c2cf99 in Read_Syscheck config/syscheck-config.c:1803
    #4 0x5612d9bf2802 in read_main_elements config/config.c:105
    #5 0x5612d9bf3f72 in ReadConfig config/config.c:279
    #6 0x5612d9bbb3e4 in Read_Syscheck_Config syscheckd/config.c:89
    #7 0x5612d9bc8b2c in main syscheckd/main.c:108
    #8 0x7fc318f99b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

SUMMARY: AddressSanitizer: 136 byte(s) leaked in 14 allocation(s).
```

**Comment by antoniomanuelfr** [comment]
Hi!!

I have been investigating the bug that @Molter73 described and I found the cause:
https://github.com/wazuh/wazuh/blob/93573876eafc2c5c38b7e4b4e5567ef3b348d9c1/src/syscheckd/run_check.c#L588-L616

When `syscheckd` enters in `fim_link_update` it checks if the new path is already added (line 614 in the above code snippet ). This check is done by comparing if there is an entry in `syscheck.dir` with the same path, and as the new folder has the same name as the link, the comparison is true, `fim_link_update` discards the link and returns.

 I'm going to do some testing here, but I think that instead of using `syscheck.dir` to check if the path is already configured, using `fim_get_real_path`  will do the trick and avoid this problem.

In the meantime, I will move this PR to draft.

Kind regards

**Comment by antoniomanuelfr** [comment]
These are the integration tests with the latest changes:
[agent_centos_html_report_test_integration_B1077_20210201104211.zip](https://github.com/wazuh/wazuh/files/5903888/agent_centos_html_report_test_integration_B1077_20210201104211.zip)
[agent_macos_html_report_test_integration_B1078_20210201104233.zip](https://github.com/wazuh/wazuh/files/5903889/agent_macos_html_report_test_integration_B1078_20210201104233.zip)
[agent_solaris_html_report_test_integration_B1078_20210201104233.zip](https://github.com/wazuh/wazuh/files/5903890/agent_solaris_html_report_test_integration_B1078_20210201104233.zip)
[agent_ubuntu_html_report_test_integration_B1077_20210201104211.zip](https://github.com/wazuh/wazuh/files/5903891/agent_ubuntu_html_report_test_integration_B1077_20210201104211.zip)
[agent_windows_html_report_test_integration_B1077_20210201104211.zip](https://github.com/wazuh/wazuh/files/5903892/agent_windows_html_report_test_integration_B1077_20210201104211.zip)

**Comment by antoniomanuelfr** [comment]
These are the integration test with the latest changes 
[agent_centos_html_report_test_integration_B1251_20210212093530.zip](https://github.com/wazuh/wazuh/files/5971550/agent_centos_html_report_test_integration_B1251_20210212093530.zip)
[agent_windows_html_report_test_integration_B1251_20210212093530.zip](https://github.com/wazuh/wazuh/files/5971554/agent_windows_html_report_test_integration_B1251_20210212093530.zip)
[agent_ubuntu_html_report_test_integration_B1251_20210212093530.zip](https://github.com/wazuh/wazuh/files/5971553/agent_ubuntu_html_report_test_integration_B1251_20210212093530.zip)
[agent_macos_html_report_test_integration_B1252_20210212093554.zip](https://github.com/wazuh/wazuh/files/5971551/agent_macos_html_report_test_integration_B1252_20210212093554.zip)
[agent_solaris_html_report_test_integration_B1252_20210212093554.zip](https://github.com/wazuh/wazuh/files/5971552/agent_solaris_html_report_test_integration_B1252_20210212093554.zip)

There is one failed test in CentOS. After looking at the logs, it seems to be a false negative. The test monitors a nonexisting folder and waits for a message that informs that the audit rule can't be created. After that, the test creates a folder and wait until the rule is added (30 seconds). The timeout between these two events is 30 seconds and in this execution, the test failed, but looking at the logs, both events appear. 

I have increased the timeout of this test to 32 seconds, in order to avoid this error in future executions.

**Review by Molter73 — APPROVED** [review]
LGTM!

**Pull request merged**